### ISSUE TITLE
Database connection slot error in Next.js 13 app directory

### ISSUE THREAD

**Issue by Eprince-hub** [vercel/next.js#45483]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:30 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T8103
    Binaries:
      Node: 19.4.0
      npm: 9.2.0
      Yarn: 1.22.19
      pnpm: 7.25.1
    Relevant packages:
      next: 13.1.6
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/autumn-mountain-0efmhn

## Run the project in `CodeSandbox`

https://codesandbox.io/p/sandbox/autumn-mountain-0efmhn

Start `editing` and `saving` any of these files => `database/connect.ts` / `app/page.tsx`

Check the console and watch the database connection count increase as the changes are saved

---

## Run the project locally

Clone Repository
```
git clone https://github.com/Eprince-hub/postgres-connection-slot-error-in-next-13-app-directory
```
Install Packages
```
yarn
```

Create a `.env` file with the following credentials and replace `xxxxxxx...` with a real database connection
```
PGHOST=localhost
PGUSERNAME=xxxxxxxxxxxxxxx
PGPASSWORD=xxxxxxxxxxxxxxx
PGDATABASE=xxxxxxxxxxxxxxx
```

Run Migration
```
yarn migrate up
```
Start `editing` and `saving` any of these files => `database/connect.ts` / `app/page.tsx`

Check the `console` and watch the database connection count increases as the changes are saved, and at a certain count, the error will happen

### Describe the Bug

With Fast Refresh (hot reloading) during the development mode, a new database connection is initiated on every refresh, Thereby causing a database connection slot error like `remaining connection slots are reserved for non-replication superuser connections` in PostgreSQL

Screenshot of the error on the browser
![Screenshot 2023-02-01 at 10 38 04](https://user-images.githubusercontent.com/74430629/216006301-c0cbf7f4-82ce-40c2-a5ed-259c5b364fb6.png)

Screenshot of the error in CLI
![Screenshot 2023-02-01 at 11 16 44](https://user-images.githubusercontent.com/74430629/216015413-e1fc7e14-820f-4681-9a10-152af2487776.png)

When a file that imports a database code is edited and saved, the Fast Refresh makes a new connection to the database on every refresh, rapidly increasing the database connection count. The database connection slot error is triggered when the count is above 90

The videos below show the process that triggers the error and the database connection count increment

https://user-images.githubusercontent.com/74430629/216009538-f99ec951-4cc3-4df8-83b3-8db9c2e08146.mp4

There is a current discussion relating to this issue here https://github.com/vercel/next.js/discussions/26427

### Expected Behavior

It is expected that the development mode Fast Refresh (hot reloading) doesn't attempt reconnecting to the database but instead uses the previous connection.

Using the `globalThis` singleton workaround described here https://github.com/vercel/next.js/discussions/26427#discussioncomment-898067, I was able to prevent this behavior.

The video below shows that the database connection count is not increasing above 13-14 counts regardless of how many reloads happen

https://user-images.githubusercontent.com/74430629/216012463-c5f7aacd-840a-4d51-ad7c-42b90f5e18e0.MOV

**Comment by abuinitski** [comment]
Pretty important issue, +1 on it. Super important for any non-trivial apps which use any global context or services (e.g. a Message Queue consumer, or gRPC clients / Protobuf parsers).

Few extra observations:

1. Workaround has to be implemented exactly as proposed, any attempt to hide it behind any tooling breaks it. E.g. "singleton" utility that maintains a collection of singleton instances in global context and provide convenience methods to get-or-instantiate an instance - doesn't work, it's like `globalContext` is also getting reinstantiated based on some funny rules. Every singleton has to define its own extension of `globalContext` and export its own `getSingletonInstance()`-kind-of-a-function.
2. A side-effect of that - `instance of` operator doesn't work at least in dev mode. If you created a custom exception class, for example, you cannot use `x instance of MyException`, since multiple versions of the same class may be present. 
3. A singleton hidden behind such workaround cannot use any standard next methods for server functions (e.g. `cookies` etc). Next is not able to find it's async context.

**Comment by sang-au** [comment]
In AWS has the proxy feature, actually that is the connection pooling, and in the .NET core I can implement something like that. So I believe you can implement something like that or NextJs have feature that support connection pooling in a case using server action to query data in db

**Comment by eashish93** [comment]
Related to : https://github.com/vercel/next.js/issues/52165

**Comment by alexeigs** [comment]
Solution when working with supabase and drizzle: https://github.com/drizzle-team/drizzle-orm/issues/928#issuecomment-1739105895

**Comment by chimung** [comment]
The workaround solution does not work anymore

**Comment by karlhorky** [comment]
I noticed that Prisma documented this `globalThis` workaround for Next.js Fast Refresh in their official docs now too:

- https://www.prisma.io/docs/orm/more/help-and-troubleshooting/help-articles/nextjs-prisma-client-dev-practices

**Comment by zero1zero** [comment]
> The workaround solution does not work anymore

Seconded that this appears to not work anymore (14.2).  Is there really no functioning way to create a singleton object in NextJS (in dev mode)? This seems like a pretty simple use case that I'm shocked is impossible, or even requires a workaround.

**Comment by karlhorky** [comment]
@zero1zero can you provide failing demo code? It has worked for numerous apps of ours on pretty much every minor/patch version of 13, 14 and 15.

**Comment by zero1zero** [comment]
I tried implementing this on a fresh CRA app with a dead simple singleton and it seems to work there. Tried the same code in my project and it _did not_ have any effect.   Sounds like a more complex project/setup (or artifact thereof) makes this workaround ineffective.

We've been having issues with very slow dev mode reloads (or reloads when no files have changed), so this may play a factor, but may not. Seems that multiple people are experiencing similar issues with no resolution, so I'm hoping someone with more of a depth of understanding in dev mode runs into this and solves it soon.

**Comment by karlhorky** [comment]
@zero1zero without demo code showing it failing or reproduction steps, it will be difficult to reproduce your problem.

It is a common solution to the problem of database connections with hot reloading - you can see this recommended in multiple official places, apart from here.

So I'm guessing there is something else with your app code or setup that is causing the problem.

I would be interested to know more from anyone experiencing failures here, but without a reproduction there is not much I can do.

**Comment by zero1zero** [comment]
> So I'm guessing there is something else with your app code or setup that is causing the problem.

Yes, I'm suggesting that a more complex or specific setup renders this workaround ineffective.  Without knowing what it is about our app that is causing it to be ineffective makes it impossible to extract out demo code to share. I tried making a CRA app with the basic patterns and simplified structure but the workaround was effective there. 

If I eventually find the cause (or a solution) I will update this thread but this is mostly so I can share that in some cases the workaround is ineffective with the hopes that someone (or NextJS) will find a working solution for singletons in dev mode.

**Comment by karlhorky** [comment]
Ok, let us know if you can reproduce the problem in a minimal reproduction.

For now, the note that it's not working in a larger app with no additional details is unfortunately not actionable and not helpful.

**Comment by Bligoubloups** [comment]
Hi !

I've got similar issue and made a minimal reproduction of it.
https://codesandbox.io/p/devbox/kvxgj2?file=%2Fbaz.ts%3A15%2C25

The issue appears when fetching a dynamic api route (`/api/foo/[bar]`).
The global variable `baz` gets initialized everytime but actually kind of never update itself 🤯.

**Comment by karlhorky** [comment]
@Bligoubloups not understanding yet what you mean - I'm having trouble making a connection between your reproduction and the database exhaustion issue / the `globalThis` workaround.

Maybe you could edit your post above to make it clear what the connection is?

**Comment by bobosostupid-maker** [comment]
The work around is not working for dynamic api route in development mode.
When you call a dynamic api route, the global.your_instance is always undefined everytime you call it. This happens it dynamic api routes only.......

**Comment by karlhorky** [comment]
@bobosostupid-maker if you can provide a failing demo on CodeSandbox or StackBlitz, I can take a look.

The workaround shouldn't fail in that way, has never in all of our projects that we use it in.

**Comment by lxy32167** [comment]
It appears that only dynamic routing causes this issue. #74204
